src/Amazon.Extensions.CognitoAuthentication/CognitoUserAuthentication.cs

```diff
@@ -413,6 +413,10 @@ public async Task<AuthFlowResponse> StartWithRefreshTokenAuthAsync(InitiateRefre
             InitiateAuthResponse initiateResponse =
                 await Provider.InitiateAuthAsync(initiateAuthRequest).ConfigureAwait(false);
 
+            // Service does not return the refresh token. Hence, set it to the old refresh token that was used.
+            if (string.IsNullOrEmpty(initiateResponse.ChallengeName) && string.IsNullOrEmpty(initiateResponse.AuthenticationResult.RefreshToken))
```

Comment:
Why do you gate `RefreshToken` assignment with  `ChallengeName`?

Can it be simplified by just checking whether the response has `RefreshToken` or not?

Reply:
The reason is that the subsequent `UpdateSessionIfAuthenticationComplete` call updates tokens only if `string.IsNullOrEmpty(challengeName)` is `true`. Hence we should only alter RefreshToken only if this condition is also true